### PR TITLE
SegmentDirectoryLoader interface and local impl

### PR DESCRIPTION
**Pull request by npawar**
Introducing a `SegmentDirectoryLoader` which will create the `SegmentDirectory`, with default implementation being `SegmentLocalFSDirectory`.
As part of this PR, the `SegmentDirectory` had to move to `pinot-segment-spi`, and with that came `SegmentMetadataImpl`, `PinotDataBuffer`, `ColumnIndexDirectory`, `V1Constants`, `SegmentDirectoryPaths`. Here's the complete list of files that were moved from local to spi:
```
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/store/ColumnIndexDirectory.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/store/ColumnIndexType.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/store/SegmentDirectory.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/store/SegmentDirectoryPaths.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/ColumnMetadata.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java

pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/BasePinotLBuffer.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/CleanerUtil.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/PinotByteBuffer.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/PinotDataBuffer.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/PinotNativeOrderLBuffer.java
pinot-segment-spi/src/test/java/org/apache/pinot/segment/spi/memory/PinotDataBufferTest.java

pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/partition/ByteArrayPartitionFunction.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/partition/HashCodePartitionFunction.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/partition/ModuloPartitionFunction.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/partition/MurmurPartitionFunction.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/partition/PartitionFunctionFactory.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/partition/ColumnPartitionMetadata.java
pinot-segment-spi/src/test/java/org/apache/pinot/segment/spi/partition/PartitionFunctionTest.java

pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/V1Constants.java
pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/creator/TextIndexType.java
```

Only these are the files with the new changes. Rest all are simply package name changes or imports
```
ImmutableSegmentLoader.java
SegmentV1V2ToV3FormatConverter.java
IndexLoadingConfig.java
LocalSegmentDirectoryLoader.java
SegmentLocalFSDirectory.java
SegmentDirectory.java 
HelixInstanceDataManagerConfig.java
SegmentLoader.java
SegmentDirectoryLoaderRegistry.java
SegmentDirectoryLoader.java
SegmentPreProcessor.java
SegmentMetadataImpl.java 


BaseDefaultColumnHandlerTest
SegmentLocalFSDirectoryTest
LoaderTest
```

### CODE REVIEW

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/loader/SegmentDirectoryLoaderRegistry.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  private SegmentDirectoryLoaderRegistry() {
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentDirectoryLoaderRegistry.class);
+  private static final Map<String, SegmentDirectoryLoader> _segmentDirectoryLoaderMap = new HashMap<>();
+
+  static {
+    Reflections reflections = new Reflections(
+        new ConfigurationBuilder().setUrls(ClasspathHelper.forPackage("org.apache.pinot.segment"))
+            .filterInputsBy(new FilterBuilder.Include(".*\\.loader\\..*"))
```
Comment (mayankshriv) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/loader/SegmentDirectoryLoaderRegistry.java:49]: Is there tests to ensure that any future package move does not break stuff?

Comment (npawar) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/loader/SegmentDirectoryLoaderRegistry.java:49]: Yes, many of the existing tests now use the registry to get the local loader. If the loader is moved, all those tests will fail.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/loader/SegmentDirectoryLoader.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * Interface for creating and loading the {@link SegmentDirectory} instance using provided config
+ */
+public interface SegmentDirectoryLoader {
+
+  /**
+   * Creates the {@link SegmentDirectory} instance
+   * @param indexDir index directory
+   * @param config config for SegmentDirectory
+   */
+  SegmentDirectory load(URI indexDir, PinotConfiguration config)
```
Comment (kishoreg) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/loader/SegmentDirectoryLoader.java:36]: what will this config contain.. will it be specific to the backend or entire server configuration?

**Review by kishoreg — APPROVED** [review]

File: pinot-broker/src/test/java/org/apache/pinot/broker/routing/segmentpruner/SegmentPrunerTest.java [hunk 3]
```diff
@@ -34,7 +34,7 @@
 import org.apache.helix.model.IdealState;
 import org.apache.helix.store.zk.ZkHelixPropertyStore;
 import org.apache.pinot.common.metadata.ZKMetadataProvider;
-import org.apache.pinot.common.metadata.segment.ColumnPartitionMetadata;
+import org.apache.pinot.segment.spi.partition.ColumnPartitionMetadata;
```
[the remark this answers is not in the document]
Comment (npawar) [reply pinot-broker/src/test/java/org/apache/pinot/broker/routing/segmentpruner/SegmentPrunerTest.java]: I didn't create a new one. It already existed. `PartitionFunction` was in the package. It also was the logical place for these classes to move, which used to be in `org.apache.pinot.segment.local.partition`
```
MurmurPartitionFunction.java
ModuloPartitionFunction.java
HashPartitionFunction.java
ByteArrayPartitionFunction.java
PartitionFunctionFactory.java
```
The ColumnPartitionMetadata is also only relevant to partition. I can put it in `metadata` package under this if that's better.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/loader/SegmentDirectoryLoader.java (hunk 2 above)
Comment (npawar) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/loader/SegmentDirectoryLoader.java:36]: Specific to tier backend. WIll rename the config

**Pull request merged**